<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (1.8.0_162) on Tue Mar 27 17:27:36 CEST 2018 -->
<title>MEReactionPlugin</title>
<meta name="date" content="2018-03-27">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="MEReactionPlugin";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":10,"i13":10,"i14":10,"i15":10,"i16":10,"i17":10,"i18":10,"i19":10,"i20":10,"i21":10,"i22":10,"i23":10,"i24":10,"i25":10,"i26":10,"i27":10,"i28":10,"i29":10,"i30":10,"i31":10,"i32":10,"i33":10,"i34":10,"i35":10,"i36":10,"i37":10,"i38":10,"i39":10,"i40":10,"i41":10,"i42":10,"i43":10,"i44":10,"i45":10,"i46":10,"i47":10,"i48":10,"i49":10,"i50":10,"i51":10,"i52":10,"i53":10,"i54":10,"i55":10,"i56":10,"i57":10,"i58":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MEReactionPlugin.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../sbmlme/MEProcessData.html" title="class in sbmlme"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../sbmlme/MESBOLPlugin.html" title="class in sbmlme"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?sbmlme/MEReactionPlugin.html" target="_top">Frames</a></li>
<li><a href="MEReactionPlugin.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">sbmlme</div>
<h2 title="Class MEReactionPlugin" class="title">Class MEReactionPlugin</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.AbstractTreeNode</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.xml.XMLToken</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.xml.XMLNode</li>
<li>
<ul class="inheritance">
<li><a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">sbmlme.MEAbstractXMLNodePlugin</a></li>
<li>
<ul class="inheritance">
<li>sbmlme.MEReactionPlugin</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable, java.lang.Cloneable, javax.swing.tree.TreeNode, org.sbml.jsbml.util.TreeNodeWithChangeSupport, <a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">MEReactionPlugin</span>
extends <a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">MEAbstractXMLNodePlugin</a>
implements <a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></pre>
<div class="block">Implements attributes and methods to encode all information of reactions in a
 COBRAme model in SBML.
 <p>
 Besides SBML core the packages fbc and groups are used in all reaction
 creating methods of this class. The groups package is used to encode the
 reaction type. The fbc package is used to encode the upper and lower bounds
 and the objective coefficient of the reaction.
 </p></div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Marc A. Voigt</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../serialized-form.html#sbmlme.MEReactionPlugin">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>private static long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#serialVersionUID">serialVersionUID</a></span></code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.org.sbml.jsbml.xml.XMLToken">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;org.sbml.jsbml.xml.XMLToken</h3>
<code>attributes, characters, column, isEndElement, isEOF, isStartElement, isText, line, namespaces, triple</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.org.sbml.jsbml.AbstractTreeNode">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;org.sbml.jsbml.AbstractTreeNode</h3>
<code>listOfListeners, parent, resourceBundle</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.sbmlme.MEConstants">
<!--   -->
</a>
<h3>Fields inherited from interface&nbsp;sbmlme.<a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></h3>
<code><a href="../sbmlme/MEConstants.html#aggregationPropensity">aggregationPropensity</a>, <a href="../sbmlme/MEConstants.html#aminoAcid">aminoAcid</a>, <a href="../sbmlme/MEConstants.html#biomassType">biomassType</a>, <a href="../sbmlme/MEConstants.html#codon">codon</a>, <a href="../sbmlme/MEConstants.html#coefficient">coefficient</a>, <a href="../sbmlme/MEConstants.html#complexFormationReaction">complexFormationReaction</a>, <a href="../sbmlme/MEConstants.html#complexId">complexId</a>, <a href="../sbmlme/MEConstants.html#dataId">dataId</a>, <a href="../sbmlme/MEConstants.html#dna">dna</a>, <a href="../sbmlme/MEConstants.html#dnaSegment">dnaSegment</a>, <a href="../sbmlme/MEConstants.html#element">element</a>, <a href="../sbmlme/MEConstants.html#enzymeRef">enzymeRef</a>, <a href="../sbmlme/MEConstants.html#fixedkeff">fixedkeff</a>, <a href="../sbmlme/MEConstants.html#genericFormation">genericFormation</a>, <a href="../sbmlme/MEConstants.html#genomePos">genomePos</a>, <a href="../sbmlme/MEConstants.html#id">id</a>, <a href="../sbmlme/MEConstants.html#keff">keff</a>, <a href="../sbmlme/MEConstants.html#key">key</a>, <a href="../sbmlme/MEConstants.html#lengthDependent">lengthDependent</a>, <a href="../sbmlme/MEConstants.html#lowerFluxBound">lowerFluxBound</a>, <a href="../sbmlme/MEConstants.html#meReaction">meReaction</a>, <a href="../sbmlme/MEConstants.html#messengerRNA">messengerRNA</a>, <a href="../sbmlme/MEConstants.html#metabolicReaction">metabolicReaction</a>, <a href="../sbmlme/MEConstants.html#mRNA">mRNA</a>, <a href="../sbmlme/MEConstants.html#multiplier">multiplier</a>, <a href="../sbmlme/MEConstants.html#name">name</a>, <a href="../sbmlme/MEConstants.html#ncRNA">ncRNA</a>, <a href="../sbmlme/MEConstants.html#noncodingRNA">noncodingRNA</a>, <a href="../sbmlme/MEConstants.html#ns">ns</a>, <a href="../sbmlme/MEConstants.html#postTranslationReaction">postTranslationReaction</a>, <a href="../sbmlme/MEConstants.html#prefix">prefix</a>, <a href="../sbmlme/MEConstants.html#processed">processed</a>, <a href="../sbmlme/MEConstants.html#propensityScaling">propensityScaling</a>, <a href="../sbmlme/MEConstants.html#protein">protein</a>, <a href="../sbmlme/MEConstants.html#proteinComplex">proteinComplex</a>, <a href="../sbmlme/MEConstants.html#rate">rate</a>, <a href="../sbmlme/MEConstants.html#reactionPlugin">reactionPlugin</a>, <a href="../sbmlme/MEConstants.html#reverse">reverse</a>, <a href="../sbmlme/MEConstants.html#ribosomalRNA">ribosomalRNA</a>, <a href="../sbmlme/MEConstants.html#rRNA">rRNA</a>, <a href="../sbmlme/MEConstants.html#sbolAnnotation">sbolAnnotation</a>, <a href="../sbmlme/MEConstants.html#sbolLoc">sbolLoc</a>, <a href="../sbmlme/MEConstants.html#sbolURI">sbolURI</a>, <a href="../sbmlme/MEConstants.html#seq">seq</a>, <a href="../sbmlme/MEConstants.html#sequence">sequence</a>, <a href="../sbmlme/MEConstants.html#species">species</a>, <a href="../sbmlme/MEConstants.html#speciesPlugin">speciesPlugin</a>, <a href="../sbmlme/MEConstants.html#speciesRef">speciesRef</a>, <a href="../sbmlme/MEConstants.html#stoichiometry">stoichiometry</a>, <a href="../sbmlme/MEConstants.html#subreaction">subreaction</a>, <a href="../sbmlme/MEConstants.html#subunit">subunit</a>, <a href="../sbmlme/MEConstants.html#summaryVariable">summaryVariable</a>, <a href="../sbmlme/MEConstants.html#surfaceAreaInner">surfaceAreaInner</a>, <a href="../sbmlme/MEConstants.html#surfaceAreaOuter">surfaceAreaOuter</a>, <a href="../sbmlme/MEConstants.html#synthetase">synthetase</a>, <a href="../sbmlme/MEConstants.html#temperature">temperature</a>, <a href="../sbmlme/MEConstants.html#transcription">transcription</a>, <a href="../sbmlme/MEConstants.html#transferRNA">transferRNA</a>, <a href="../sbmlme/MEConstants.html#translation">translation</a>, <a href="../sbmlme/MEConstants.html#translocation">translocation</a>, <a href="../sbmlme/MEConstants.html#tRNA">tRNA</a>, <a href="../sbmlme/MEConstants.html#tRNACharging">tRNACharging</a>, <a href="../sbmlme/MEConstants.html#unprocessed">unprocessed</a>, <a href="../sbmlme/MEConstants.html#upperFluxBound">upperFluxBound</a>, <a href="../sbmlme/MEConstants.html#value">value</a>, <a href="../sbmlme/MEConstants.html#variableKind">variableKind</a>, <a href="../sbmlme/MEConstants.html#versionOne">versionOne</a></code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#MEReactionPlugin--">MEReactionPlugin</a></span>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#MEReactionPlugin-sbmlme.MEReactionPlugin-">MEReactionPlugin</a></span>(<a href="../sbmlme/MEReactionPlugin.html" title="class in sbmlme">MEReactionPlugin</a>&nbsp;rp)</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#addCOBRAmeSpeciesToReaction-org.sbml.jsbml.Model-org.sbml.jsbml.Reaction-java.lang.String-java.lang.String-">addCOBRAmeSpeciesToReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                           org.sbml.jsbml.Reaction&nbsp;reaction,
                           java.lang.String&nbsp;speciesId,
                           java.lang.String&nbsp;coefficient)</code>
<div class="block">Adds a single species from a COBRAme reaction list to the corresponding
 SBML reaction.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createComplexFormationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-java.util.LinkedHashMap-">createComplexFormationReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                              org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                              org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                              java.lang.String&nbsp;id,
                              java.lang.String&nbsp;name,
                              java.lang.String&nbsp;dataId,
                              java.lang.String&nbsp;complexId,
                              double&nbsp;upperBound,
                              double&nbsp;lowerBound,
                              java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                              java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                              double&nbsp;objectiveCoefficient,
                              java.lang.String&nbsp;variableKind,
                              java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                              java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;stoichiometricMap)</code>
<div class="block">Create and add a complex formation reaction to the model.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createGenericFormationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-">createGenericFormationReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                              org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                              org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                              java.lang.String&nbsp;id,
                              java.lang.String&nbsp;name,
                              double&nbsp;upperBound,
                              double&nbsp;lowerBound,
                              java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                              java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                              double&nbsp;objectiveCoefficient,
                              java.lang.String&nbsp;variableKind)</code>
<div class="block">Create and add a formation reaction which transforms a species into a
 generic Component.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createMEReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-">createMEReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                java.lang.String&nbsp;id,
                java.lang.String&nbsp;name,
                double&nbsp;upperBound,
                double&nbsp;lowerBound,
                java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                double&nbsp;objectiveCoefficient,
                java.lang.String&nbsp;variableKind)</code>
<div class="block">Create and add an ME exclusive reaction to the model.</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createMetabolicReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-double-boolean-java.lang.String-">createMetabolicReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                       org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                       org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                       java.lang.String&nbsp;id,
                       java.lang.String&nbsp;name,
                       java.lang.String&nbsp;dataId,
                       double&nbsp;upperBound,
                       double&nbsp;lowerBound,
                       java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                       java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                       double&nbsp;objectiveCoefficient,
                       java.lang.String&nbsp;variableKind,
                       double&nbsp;keff,
                       boolean&nbsp;reverse,
                       java.lang.String&nbsp;complexDataId)</code>
<div class="block">Create and add a metabolic reaction to the model.</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createPostTranslationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-double-java.util.List-java.util.List-double-java.util.List-java.util.List-java.util.List-java.util.List-java.util.List-java.util.List-java.lang.String-java.lang.String-java.lang.String-">createPostTranslationReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                             org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                             org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                             java.lang.String&nbsp;id,
                             java.lang.String&nbsp;name,
                             java.lang.String&nbsp;dataId,
                             double&nbsp;upperBound,
                             double&nbsp;lowerBound,
                             java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                             java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                             double&nbsp;objectiveCoefficient,
                             java.lang.String&nbsp;variableKind,
                             java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                             double&nbsp;aggregationPropensity,
                             java.util.List&lt;java.lang.String&gt;&nbsp;translocation,
                             java.util.List&lt;java.lang.Double&gt;&nbsp;multipliers,
                             double&nbsp;propensityScaling,
                             java.util.List&lt;java.lang.String&gt;&nbsp;surfaceArea,
                             java.util.List&lt;java.lang.Double&gt;&nbsp;surfaceAreaValue,
                             java.util.List&lt;java.lang.String&gt;&nbsp;keqFolding,
                             java.util.List&lt;java.lang.Double&gt;&nbsp;keqValues,
                             java.util.List&lt;java.lang.String&gt;&nbsp;kFolding,
                             java.util.List&lt;java.lang.Double&gt;&nbsp;kValues,
                             java.lang.String&nbsp;biomassType,
                             java.lang.String&nbsp;processed,
                             java.lang.String&nbsp;unprocessed)</code>
<div class="block">Create and add a post translation reaction to the model.</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createSummaryVariableReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.util.List-java.util.List-double-java.lang.String-">createSummaryVariableReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                             org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                             org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                             java.lang.String&nbsp;id,
                             java.lang.String&nbsp;name,
                             java.lang.String&nbsp;upperBound,
                             java.lang.String&nbsp;lowerBound,
                             java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                             java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                             double&nbsp;objectiveCoefficient,
                             java.lang.String&nbsp;variableKind)</code>
<div class="block">Create and add a SummaryVariable Reaction to the model.</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createTranscriptionReaction-org.sbml.jsbml.Model-org.sbolstandard.core2.SBOLDocument-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.lang.String-java.util.LinkedHashMap-">createTranscriptionReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                           org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                           org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                           org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                           java.lang.String&nbsp;id,
                           java.lang.String&nbsp;name,
                           java.lang.String&nbsp;dataId,
                           double&nbsp;upperBound,
                           double&nbsp;lowerBound,
                           java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                           java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                           double&nbsp;objectiveCoefficient,
                           java.lang.String&nbsp;variableKind,
                           java.lang.String&nbsp;sequence,
                           java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap)</code>
<div class="block">Create and add a transcription reaction to the model.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createTranslationReaction-org.sbml.jsbml.Model-org.sbolstandard.core2.SBOLDocument-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.lang.String-java.util.LinkedHashMap-">createTranslationReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                         org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                         org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                         org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                         java.lang.String&nbsp;id,
                         java.lang.String&nbsp;name,
                         java.lang.String&nbsp;dataId,
                         double&nbsp;upperBound,
                         double&nbsp;lowerBound,
                         java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                         java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                         double&nbsp;objectiveCoefficient,
                         java.lang.String&nbsp;variableKind,
                         java.lang.String&nbsp;sequence,
                         java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap)</code>
<div class="block">Create and add a translation reaction to the model.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#createtRNAChargingReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-java.lang.String-java.lang.String-java.lang.String-">createtRNAChargingReaction</a></span>(org.sbml.jsbml.Model&nbsp;model,
                          org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                          org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                          java.lang.String&nbsp;id,
                          java.lang.String&nbsp;name,
                          java.lang.String&nbsp;dataId,
                          double&nbsp;upperBound,
                          double&nbsp;lowerBound,
                          double&nbsp;keff,
                          java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                          java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                          double&nbsp;objectiveCoefficient,
                          java.lang.String&nbsp;variableKind,
                          java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                          java.lang.String&nbsp;synthetase,
                          java.lang.String&nbsp;codon,
                          java.lang.String&nbsp;aminoAcid)</code>
<div class="block">Create and add a tRNA charging reaction to the model.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getAggregationPropensity--">getAggregationPropensity</a></span>()</code>
<div class="block">Returns the value of the attribute "aggregationPropensity".</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getAminoAcid--">getAminoAcid</a></span>()</code>
<div class="block">Returns the value of the attribute "aminoAcid".</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getBiomassType--">getBiomassType</a></span>()</code>
<div class="block">Returns the value of the attribute "biomassType".</div>
</td>
</tr>
<tr id="i13" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getCodon--">getCodon</a></span>()</code>
<div class="block">Returns the value of the attribute "codon".</div>
</td>
</tr>
<tr id="i14" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getComplexId--">getComplexId</a></span>()</code>
<div class="block">Returns the value of the attribute "complexId".</div>
</td>
</tr>
<tr id="i15" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getDataId--">getDataId</a></span>()</code>
<div class="block">Returns the value of the attribute "dataId".</div>
</td>
</tr>
<tr id="i16" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getKeff--">getKeff</a></span>()</code>
<div class="block">Returns the value of the attribute "keff".</div>
</td>
</tr>
<tr id="i17" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getProcessed--">getProcessed</a></span>()</code>
<div class="block">Returns the value of the attribute "processedProteinId".</div>
</td>
</tr>
<tr id="i18" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getPropensityScaling--">getPropensityScaling</a></span>()</code>
<div class="block">Returns the value of the attribute "propensityScaling".</div>
</td>
</tr>
<tr id="i19" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getReverse--">getReverse</a></span>()</code>
<div class="block">Returns the value of the attribute "reverse".</div>
</td>
</tr>
<tr id="i20" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getSequence--">getSequence</a></span>()</code>
<div class="block">Returns the value of the attribute "sequence".</div>
</td>
</tr>
<tr id="i21" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getSurfaceAreaInner--">getSurfaceAreaInner</a></span>()</code>
<div class="block">Returns the value of the attribute "surfaceAreaInnerMembrane".</div>
</td>
</tr>
<tr id="i22" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getSurfaceAreaOuter--">getSurfaceAreaOuter</a></span>()</code>
<div class="block">Returns the value of the attribute "surfaceAreaOuterMembrane".</div>
</td>
</tr>
<tr id="i23" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getSynthetase--">getSynthetase</a></span>()</code>
<div class="block">Returns the value of the attribute "synthetase".</div>
</td>
</tr>
<tr id="i24" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getUnprocessed--">getUnprocessed</a></span>()</code>
<div class="block">Returns the value of the attribute "unprocessedProteinId".</div>
</td>
</tr>
<tr id="i25" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#getVariableKind--">getVariableKind</a></span>()</code>
<div class="block">Returns the value of the attribute "variableKind".</div>
</td>
</tr>
<tr id="i26" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetAggregationPropensity--">isSetAggregationPropensity</a></span>()</code>
<div class="block">Returns whether the attribute "aggregationPropensity" is set.</div>
</td>
</tr>
<tr id="i27" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetAminoAcid--">isSetAminoAcid</a></span>()</code>
<div class="block">Returns whether the attribute "aminoAcid" is set.</div>
</td>
</tr>
<tr id="i28" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetBiomassType--">isSetBiomassType</a></span>()</code>
<div class="block">Returns whether the attribute "biomassType" is set.</div>
</td>
</tr>
<tr id="i29" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetCodon--">isSetCodon</a></span>()</code>
<div class="block">Returns whether the attribute "codon" is set.</div>
</td>
</tr>
<tr id="i30" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetComplexId--">isSetComplexId</a></span>()</code>
<div class="block">Returns whether the attribute "complexId" is set.</div>
</td>
</tr>
<tr id="i31" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetDataId--">isSetDataId</a></span>()</code>
<div class="block">Returns whether the attribute "dataId" is set.</div>
</td>
</tr>
<tr id="i32" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetKeff--">isSetKeff</a></span>()</code>
<div class="block">Returns whether the attribute "keff" is set.</div>
</td>
</tr>
<tr id="i33" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetProcessed--">isSetProcessed</a></span>()</code>
<div class="block">Returns whether the attribute "processedProteinId" is set.</div>
</td>
</tr>
<tr id="i34" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetPropensityScaling--">isSetPropensityScaling</a></span>()</code>
<div class="block">Returns whether the attribute "propensityScaling" is set.</div>
</td>
</tr>
<tr id="i35" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetReverse--">isSetReverse</a></span>()</code>
<div class="block">Returns whether the attribute "reverse" is set.</div>
</td>
</tr>
<tr id="i36" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetSequence--">isSetSequence</a></span>()</code>
<div class="block">Returns whether the attribute "sequence" is set.</div>
</td>
</tr>
<tr id="i37" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetSurfaceAreaInner--">isSetSurfaceAreaInner</a></span>()</code>
<div class="block">Returns whether the attribute "surfaceAreaInnerMembrane" is set.</div>
</td>
</tr>
<tr id="i38" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetSurfaceAreaOuter--">isSetSurfaceAreaOuter</a></span>()</code>
<div class="block">Returns whether the attribute "surfaceAreaOuterMembrane" is set.</div>
</td>
</tr>
<tr id="i39" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetSynthetase--">isSetSynthetase</a></span>()</code>
<div class="block">Returns whether the attribute "synthetase" is set.</div>
</td>
</tr>
<tr id="i40" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetUnprocessed--">isSetUnprocessed</a></span>()</code>
<div class="block">Returns whether the attribute "unprocessedProteinId" is set.</div>
</td>
</tr>
<tr id="i41" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#isSetVariableKind--">isSetVariableKind</a></span>()</code>
<div class="block">Returns whether the attribute "variableKind" is set.</div>
</td>
</tr>
<tr id="i42" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setAggregationPropensity-java.lang.String-">setAggregationPropensity</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "aggregationPropensity".</div>
</td>
</tr>
<tr id="i43" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setAminoAcid-java.lang.String-">setAminoAcid</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "aminoAcid".</div>
</td>
</tr>
<tr id="i44" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setBiomassType-java.lang.String-">setBiomassType</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "biomassType".</div>
</td>
</tr>
<tr id="i45" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setBounds-org.sbml.jsbml.Model-org.sbml.jsbml.ext.fbc.FBCReactionPlugin-java.lang.String-java.lang.String-">setBounds</a></span>(org.sbml.jsbml.Model&nbsp;model,
         org.sbml.jsbml.ext.fbc.FBCReactionPlugin&nbsp;fbcTempReaction,
         java.lang.String&nbsp;upperBound,
         java.lang.String&nbsp;lowerBound)</code>
<div class="block">Sets the Bounds of a fbc reaction to a valid parameter.</div>
</td>
</tr>
<tr id="i46" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setCodon-java.lang.String-">setCodon</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "codon".</div>
</td>
</tr>
<tr id="i47" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setComplexId-java.lang.String-">setComplexId</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "complexId".</div>
</td>
</tr>
<tr id="i48" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setDataId-java.lang.String-">setDataId</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "dataId".</div>
</td>
</tr>
<tr id="i49" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setKeff-java.lang.String-">setKeff</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "keff".</div>
</td>
</tr>
<tr id="i50" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setProcessed-java.lang.String-">setProcessed</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "processedProteinId".</div>
</td>
</tr>
<tr id="i51" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setPropensityScaling-java.lang.String-">setPropensityScaling</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "propensityScaling".</div>
</td>
</tr>
<tr id="i52" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setReverse-java.lang.String-">setReverse</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "reverse".</div>
</td>
</tr>
<tr id="i53" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setSequence-java.lang.String-">setSequence</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "sequence".</div>
</td>
</tr>
<tr id="i54" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setSurfaceAreaInner-java.lang.String-">setSurfaceAreaInner</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "surfaceAreaInnerMembrane".</div>
</td>
</tr>
<tr id="i55" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setSurfaceAreaOuter-java.lang.String-">setSurfaceAreaOuter</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "surfaceAreaOuterMembrane".</div>
</td>
</tr>
<tr id="i56" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setSynthetase-java.lang.String-">setSynthetase</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "synthetase".</div>
</td>
</tr>
<tr id="i57" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setUnprocessed-java.lang.String-">setUnprocessed</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "unprocessedProteinId".</div>
</td>
</tr>
<tr id="i58" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MEReactionPlugin.html#setVariableKind-java.lang.String-">setVariableKind</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the mandatory attribute "variableKind".</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.sbmlme.MEAbstractXMLNodePlugin">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;sbmlme.<a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">MEAbstractXMLNodePlugin</a></h3>
<code><a href="../sbmlme/MEAbstractXMLNodePlugin.html#createSBMLConformId-java.lang.String-">createSBMLConformId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getAttribute-java.lang.String-">getAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getId--">getId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getName--">getName</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetAttribute-java.lang.String-">isSetAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetId--">isSetId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetName--">isSetName</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setAttribute-java.lang.String-java.lang.String-">setAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setId-java.lang.String-">setId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setName-java.lang.String-">setName</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.xml.XMLNode">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.xml.XMLNode</h3>
<code>addChild, clone, convertStringToXMLNode, convertStringToXMLNode, convertXMLNodeToString, getAllowsChildren, getChild, getChildAt, getChildCount, getChildElement, getChildElements, insertChild, removeChild, removeChild, removeChildren, toString, toXMLString</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.xml.XMLToken">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.xml.XMLToken</h3>
<code>addAttr, addAttr, addAttr, addAttr, addNamespace, addNamespace, append, clearAttributes, clearNamespaces, equals, getAttributes, getAttributesLength, getAttrIndex, getAttrIndex, getAttrIndex, getAttrName, getAttrPrefix, getAttrPrefixedName, getAttrURI, getAttrValue, getAttrValue, getAttrValue, getAttrValue, getCharacters, getColumn, getLine, getNamespaceIndex, getNamespaceIndexByPrefix, getNamespacePrefix, getNamespacePrefix, getNamespaces, getNamespacesLength, getNamespaceURI, getNamespaceURI, getNamespaceURI, getPrefix, getURI, hasAttr, hasAttr, hasAttr, hasAttr, hashCode, hasNamespaceNS, hasNamespacePrefix, hasNamespaceURI, isAttributesEmpty, isElement, isEnd, isEndFor, isEOF, isNamespacesEmpty, isStart, isText, removeAttr, removeAttr, removeAttr, removeAttr, removeNamespace, removeNamespace, setAttributes, setCharacters, setEnd, setEOF, setNamespaces, setTriple, unsetEnd</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.AbstractTreeNode">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.AbstractTreeNode</h3>
<code>addAllChangeListeners, addAllChangeListeners, addTreeNodeChangeListener, addTreeNodeChangeListener, children, clearUserObjects, containsUserObjectKey, filter, filter, filter, fireNodeAddedEvent, fireNodeRemovedEvent, firePropertyChange, getIndex, getListOfTreeNodeChangeListeners, getNumChildren, getParent, getRoot, getTreeNodeChangeListenerCount, getUserObject, indexOf, isCloningInProgress, isInvalidSBMLAllowed, isLeaf, isReadingInProgress, isRoot, isSetParent, isSetUserObjects, notifyChildChange, putUserObject, removeAllTreeNodeChangeListeners, removeAllTreeNodeChangeListeners, removeFromParent, removeTreeNodeChangeListener, removeTreeNodeChangeListener, removeUserObject, setParent, userObjectKeySet</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>finalize, getClass, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="serialVersionUID">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>serialVersionUID</h4>
<pre>private static final&nbsp;long serialVersionUID</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../constant-values.html#sbmlme.MEReactionPlugin.serialVersionUID">Constant Field Values</a></dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="MEReactionPlugin--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MEReactionPlugin</h4>
<pre>public&nbsp;MEReactionPlugin()</pre>
</li>
</ul>
<a name="MEReactionPlugin-sbmlme.MEReactionPlugin-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MEReactionPlugin</h4>
<pre>public&nbsp;MEReactionPlugin(<a href="../sbmlme/MEReactionPlugin.html" title="class in sbmlme">MEReactionPlugin</a>&nbsp;rp)</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="addCOBRAmeSpeciesToReaction-org.sbml.jsbml.Model-org.sbml.jsbml.Reaction-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>addCOBRAmeSpeciesToReaction</h4>
<pre>public&nbsp;void&nbsp;addCOBRAmeSpeciesToReaction(org.sbml.jsbml.Model&nbsp;model,
                                        org.sbml.jsbml.Reaction&nbsp;reaction,
                                        java.lang.String&nbsp;speciesId,
                                        java.lang.String&nbsp;coefficient)
                                 throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Adds a single species from a COBRAme reaction list to the corresponding
 SBML reaction.
 <p>
 In COBRAme reactions only possess one list for both reactants and products.
 When species are added from this list reactants and products need to be
 separated, the method does this by checking if the species' coefficient is
 negative.
 In COBRAme coefficients can also be complex formulas that depend on
 parameters. When such a coefficient is added the formula is checked for
 currently unknown parameters that need to be added to the ListOfParameters
 of the model. The check whether a species is a product or reactant is based
 on the first character in the string that is not an opening bracket, if
 this is a minus the species is assumed to be a reactant.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>reaction</code> - the SBML reaction that the species should be added to</dd>
<dd><code>speciesId</code> - the id of the species</dd>
<dd><code>coefficient</code> - the coefficient of the species</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="setBounds-org.sbml.jsbml.Model-org.sbml.jsbml.ext.fbc.FBCReactionPlugin-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setBounds</h4>
<pre>public&nbsp;void&nbsp;setBounds(org.sbml.jsbml.Model&nbsp;model,
                      org.sbml.jsbml.ext.fbc.FBCReactionPlugin&nbsp;fbcTempReaction,
                      java.lang.String&nbsp;upperBound,
                      java.lang.String&nbsp;lowerBound)
               throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Sets the Bounds of a fbc reaction to a valid parameter. If there is no
 valid
 parameter a new one is created.
 <p>
 In order to not create several parameters with the same value this method
 translates the value into a String that can be used as a valid SBML id. The
 method then searches the ListOfParameter for a parameter with the
 translated id. If no such parameter is found a new parameter is created.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>fbcTempReaction</code> - the fbc plugin for the reaction</dd>
<dd><code>upperBound</code> - the designated upper bound</dd>
<dd><code>lowerBound</code> - the designated lower bound</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createSummaryVariableReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.util.List-java.util.List-double-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createSummaryVariableReaction</h4>
<pre>public&nbsp;void&nbsp;createSummaryVariableReaction(org.sbml.jsbml.Model&nbsp;model,
                                          org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                          org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                          java.lang.String&nbsp;id,
                                          java.lang.String&nbsp;name,
                                          java.lang.String&nbsp;upperBound,
                                          java.lang.String&nbsp;lowerBound,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                          double&nbsp;objectiveCoefficient,
                                          java.lang.String&nbsp;variableKind)
                                   throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add a SummaryVariable Reaction to the model.
 <p>
 In COBRAme this reaction type is used to apply global constraints to the
 model. This method adds the SBMLme exclusive attribute "variableKind" to
 the annotation of the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createGenericFormationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createGenericFormationReaction</h4>
<pre>public&nbsp;void&nbsp;createGenericFormationReaction(org.sbml.jsbml.Model&nbsp;model,
                                           org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                           org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                           java.lang.String&nbsp;id,
                                           java.lang.String&nbsp;name,
                                           double&nbsp;upperBound,
                                           double&nbsp;lowerBound,
                                           java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                           java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                           double&nbsp;objectiveCoefficient,
                                           java.lang.String&nbsp;variableKind)
                                    throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add a formation reaction which transforms a species into a
 generic Component.
 <p>
 This method also adds the reactant to the group of the generic component.
 This method adds the SBMLme exclusive attribute "variableKind" to
 the annotation of the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createTranscriptionReaction-org.sbml.jsbml.Model-org.sbolstandard.core2.SBOLDocument-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.lang.String-java.util.LinkedHashMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createTranscriptionReaction</h4>
<pre>public&nbsp;void&nbsp;createTranscriptionReaction(org.sbml.jsbml.Model&nbsp;model,
                                        org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                                        org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                        org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                        java.lang.String&nbsp;id,
                                        java.lang.String&nbsp;name,
                                        java.lang.String&nbsp;dataId,
                                        double&nbsp;upperBound,
                                        double&nbsp;lowerBound,
                                        java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                        java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                        double&nbsp;objectiveCoefficient,
                                        java.lang.String&nbsp;variableKind,
                                        java.lang.String&nbsp;sequence,
                                        java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap)
                                 throws org.sbml.jsbml.text.parser.ParseException,
                                        org.sbolstandard.core2.SBOLValidationException</pre>
<div class="block">Create and add a transcription reaction to the model.
 <p>
 This method also creates a new ComponentDefinition with the given sequence
 in the SBOL document for the transcription unit. The method adds four
 SBMLme attributes to the annotation of the reaction, "variableKind",
 "dataId", "sequence" and "listOfSubreactionReferences". The attribute
 "sequence" contains the URI of the ComponentDefinition in the SBOL document
 for the transcription unit. The "listOfSubreactionReferences" contains all
 subreactions that take part in the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>sbol</code> - the SBOL document to which the sequence should be added</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>sequence</code> - the nucleotide sequence of the transcription unit, cannot be null or
        empty string</dd>
<dd><code>subreactionMap</code> - the map with the ids and coefficients of the subreactions that take
        part in the reaction, may be null</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
<dd><code>org.sbolstandard.core2.SBOLValidationException</code></dd>
</dl>
</li>
</ul>
<a name="createTranslationReaction-org.sbml.jsbml.Model-org.sbolstandard.core2.SBOLDocument-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.lang.String-java.util.LinkedHashMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createTranslationReaction</h4>
<pre>public&nbsp;void&nbsp;createTranslationReaction(org.sbml.jsbml.Model&nbsp;model,
                                      org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                                      org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                      org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                      java.lang.String&nbsp;id,
                                      java.lang.String&nbsp;name,
                                      java.lang.String&nbsp;dataId,
                                      double&nbsp;upperBound,
                                      double&nbsp;lowerBound,
                                      java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                      java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                      double&nbsp;objectiveCoefficient,
                                      java.lang.String&nbsp;variableKind,
                                      java.lang.String&nbsp;sequence,
                                      java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap)
                               throws org.sbml.jsbml.text.parser.ParseException,
                                      org.sbolstandard.core2.SBOLValidationException</pre>
<div class="block">Create and add a translation reaction to the model.
 <p>
 This method also creates a new ComponentDefinition with the given sequence
 in the SBOL document for the transcribed mRNA. The method adds four
 SBMLme attributes to the annotation of the reaction, "variableKind",
 "dataId", "sequence" and "listOfSubreactionReferences". The attribute
 "sequence" contains the URI of the ComponentDefinition in the SBOL document
 for the transcription unit. The "listOfSubreactionReferences" contains all
 subreactions that take part in the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>sbol</code> - the SBOL document to which the sequence should be added</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>sequence</code> - the nucleotide sequence of the transcribed mRNA, cannot be null or
        empty string</dd>
<dd><code>subreactionMap</code> - the map with the ids and coefficients of the subreactions that take
        part in the reaction, may be null</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
<dd><code>org.sbolstandard.core2.SBOLValidationException</code></dd>
</dl>
</li>
</ul>
<a name="createtRNAChargingReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-java.lang.String-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createtRNAChargingReaction</h4>
<pre>public&nbsp;void&nbsp;createtRNAChargingReaction(org.sbml.jsbml.Model&nbsp;model,
                                       org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                       org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                       java.lang.String&nbsp;id,
                                       java.lang.String&nbsp;name,
                                       java.lang.String&nbsp;dataId,
                                       double&nbsp;upperBound,
                                       double&nbsp;lowerBound,
                                       double&nbsp;keff,
                                       java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                       java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                       double&nbsp;objectiveCoefficient,
                                       java.lang.String&nbsp;variableKind,
                                       java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                                       java.lang.String&nbsp;synthetase,
                                       java.lang.String&nbsp;codon,
                                       java.lang.String&nbsp;aminoAcid)
                                throws org.sbml.jsbml.text.parser.ParseException,
                                       org.sbolstandard.core2.SBOLValidationException</pre>
<div class="block">Create and add a tRNA charging reaction to the model.
 <p>
 The method adds the attributes "dataId", "keff", "variableKind",
 "synthetase", "codon", "aminoAcid" and "listOfSubreactionReferences" to the
 annotation of the reaction. The "listOfSubreactionReferences" contains all
 subreactions that take part in the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>keff</code> - the effective turnover rate of the synthetase of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>subreactionMap</code> - the map with the ids and coefficients of the subreactions that take
        part in the reaction, may be null</dd>
<dd><code>synthetase</code> - the id of the tRNA synthetase to charge the tRNA with an amino acid</dd>
<dd><code>codon</code> - the codon that is translated by the tRNA</dd>
<dd><code>aminoAcid</code> - the id of the amino acid that the tRNA transfers to a peptide</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
<dd><code>org.sbolstandard.core2.SBOLValidationException</code></dd>
</dl>
</li>
</ul>
<a name="createPostTranslationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-double-java.util.List-java.util.List-double-java.util.List-java.util.List-java.util.List-java.util.List-java.util.List-java.util.List-java.lang.String-java.lang.String-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createPostTranslationReaction</h4>
<pre>public&nbsp;void&nbsp;createPostTranslationReaction(org.sbml.jsbml.Model&nbsp;model,
                                          org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                          org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                          java.lang.String&nbsp;id,
                                          java.lang.String&nbsp;name,
                                          java.lang.String&nbsp;dataId,
                                          double&nbsp;upperBound,
                                          double&nbsp;lowerBound,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                          double&nbsp;objectiveCoefficient,
                                          java.lang.String&nbsp;variableKind,
                                          java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                                          double&nbsp;aggregationPropensity,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;translocation,
                                          java.util.List&lt;java.lang.Double&gt;&nbsp;multipliers,
                                          double&nbsp;propensityScaling,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;surfaceArea,
                                          java.util.List&lt;java.lang.Double&gt;&nbsp;surfaceAreaValue,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;keqFolding,
                                          java.util.List&lt;java.lang.Double&gt;&nbsp;keqValues,
                                          java.util.List&lt;java.lang.String&gt;&nbsp;kFolding,
                                          java.util.List&lt;java.lang.Double&gt;&nbsp;kValues,
                                          java.lang.String&nbsp;biomassType,
                                          java.lang.String&nbsp;processed,
                                          java.lang.String&nbsp;unprocessed)
                                   throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add a post translation reaction to the model.
 <p>
 The method adds the SBMLme attributes "dataID", "variableKind",
 "aggregationPropensity", "propensityScaling", "biomassType",
 "surfaceAreaInnerMembrane", "surfaceAreaOuterMembrane",
 "listOfSubreactionReferences" and "listOfTranslocationReferences" to the
 annotation of the reaction. The "listOfSubreactionReferences" contains all
 "subreactionReference" that take part in the reaction. The
 "listOfTranslocationReferences" contains all "translocationReferences" that
 take part in the reaction. The attributes "surfaceAreaInnerMembrane" and
 "surfaceAreaOuterMembrane" contain the value of the area occupied by the
 protein in the inner/outer membrane.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>subreactionMap</code> - the map with the ids and coefficients of the subreactions that take
        part in the reaction, may be null</dd>
<dd><code>aggregationPropensity</code> - the aggregation propensity for the protein</dd>
<dd><code>translocation</code> - the list of ids of translocation pathways involved in the reaction,
        may be null</dd>
<dd><code>multipliers</code> - the list of multipliers belonging to the translocations, the order
        of both lists must be the same, may be null</dd>
<dd><code>propensityScaling</code> - accounts for the propensity of some peptides to be folded by certain
        chaperones</dd>
<dd><code>surfaceArea</code> - the list for the possible surfaces, valid values are
        "SA_inner_membrane" and "SA_outer_membrane", may be null</dd>
<dd><code>surfaceAreaValue</code> - the list for the area occupied by the protein in the surfaces given
        in the list "surfaceArea", the order of both lists must be the same,
        may be null</dd>
<dd><code>keqFolding</code> - list of temperatures for keq folding, may be null</dd>
<dd><code>keqValues</code> - list of temperature dependent keq folding values for the protein,
        may be null</dd>
<dd><code>kFolding</code> - list of temperatures for k folding, may be null</dd>
<dd><code>kValues</code> - list of temperature dependent rate constants for protein folding,
        may be null</dd>
<dd><code>biomassType</code> - the id of the biomass type added by the subreactions of the
        reaction, may be null</dd>
<dd><code>processed</code> - the id of the modified species</dd>
<dd><code>unprocessed</code> - the id of the species that is modified by the reaction</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createComplexFormationReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-java.util.LinkedHashMap-java.util.LinkedHashMap-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createComplexFormationReaction</h4>
<pre>public&nbsp;void&nbsp;createComplexFormationReaction(org.sbml.jsbml.Model&nbsp;model,
                                           org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                           org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                           java.lang.String&nbsp;id,
                                           java.lang.String&nbsp;name,
                                           java.lang.String&nbsp;dataId,
                                           java.lang.String&nbsp;complexId,
                                           double&nbsp;upperBound,
                                           double&nbsp;lowerBound,
                                           java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                           java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                           double&nbsp;objectiveCoefficient,
                                           java.lang.String&nbsp;variableKind,
                                           java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;subreactionMap,
                                           java.util.LinkedHashMap&lt;java.lang.String,java.lang.Double&gt;&nbsp;stoichiometricMap)
                                    throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add a complex formation reaction to the model.
 <p>
 The method adds the SBMLme attributes "dataId", "variableKind" and
 "ListOfSubreactionReferences". The "listOfSubreactionReferences" contains
 all subreactions that take part in the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>subreactionMap</code> - the map with the ids and coefficients of the subreactions that take
        part in the reaction, may be null</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createMetabolicReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-double-boolean-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createMetabolicReaction</h4>
<pre>public&nbsp;void&nbsp;createMetabolicReaction(org.sbml.jsbml.Model&nbsp;model,
                                    org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                    org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                                    java.lang.String&nbsp;id,
                                    java.lang.String&nbsp;name,
                                    java.lang.String&nbsp;dataId,
                                    double&nbsp;upperBound,
                                    double&nbsp;lowerBound,
                                    java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                                    java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                                    double&nbsp;objectiveCoefficient,
                                    java.lang.String&nbsp;variableKind,
                                    double&nbsp;keff,
                                    boolean&nbsp;reverse,
                                    java.lang.String&nbsp;complexDataId)
                             throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add a metabolic reaction to the model.
 <p>
 In COBRAme all reversible enzymatic reactions are broken into two
 irreversible reactions representing the two directions of the reaction.
 Therefore the boolean attribute "reverse" is added to the annotation of the
 reaction to define the direction. The attributes "dataId", "keff",
 "variableKind" and "complexDataId" are also added to the annotation.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>dataId</code> - the id of the process data object that is referred by the reaction
        in COBRAme, cannot be null or empty string</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>keff</code> - the effective turnover rate of the enzymes in the translocation
        pathway</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dd><code>keff</code> - the effective turnover rate that couples enzymatic dilution to
        metabolic flux</dd>
<dd><code>reverse</code> - whether the reaction corresponds to the reverse reaction of the
        reversible reaction</dd>
<dd><code>complexDataId</code> - the id of the ComplexData object that details the enzyme which
        catalyzes the reaction</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="createMEReaction-org.sbml.jsbml.Model-org.sbml.jsbml.ext.groups.GroupsModelPlugin-org.sbml.jsbml.ext.fbc.Objective-java.lang.String-java.lang.String-double-double-java.util.List-java.util.List-double-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createMEReaction</h4>
<pre>public&nbsp;void&nbsp;createMEReaction(org.sbml.jsbml.Model&nbsp;model,
                             org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                             org.sbml.jsbml.ext.fbc.Objective&nbsp;objective,
                             java.lang.String&nbsp;id,
                             java.lang.String&nbsp;name,
                             double&nbsp;upperBound,
                             double&nbsp;lowerBound,
                             java.util.List&lt;java.lang.String&gt;&nbsp;speciesIds,
                             java.util.List&lt;java.lang.String&gt;&nbsp;coefficients,
                             double&nbsp;objectiveCoefficient,
                             java.lang.String&nbsp;variableKind)
                      throws org.sbml.jsbml.text.parser.ParseException</pre>
<div class="block">Create and add an ME exclusive reaction to the model.
 <p>
 This method adds the SBMLme exclusive attribute "variableKind" to the
 annotation of the reaction.</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>objective</code> - the active objective of the SBML model</dd>
<dd><code>id</code> - the id of the reaction to be added</dd>
<dd><code>name</code> - the name of the reaction, can be null</dd>
<dd><code>upperBound</code> - the upper bound of the reaction</dd>
<dd><code>lowerBound</code> - the lower bound of the reaction</dd>
<dd><code>speciesIds</code> - the list with the ids of all species in the reaction</dd>
<dd><code>coefficients</code> - the list with all species coefficients in the reaction, must have
        the same order as the list with species ids</dd>
<dd><code>objectiveCoefficient</code> - the coefficient of the flux objective of the reaction</dd>
<dd><code>variableKind</code> - can be either "continuous" or "discrete"</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbml.jsbml.text.parser.ParseException</code></dd>
</dl>
</li>
</ul>
<a name="isSetSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetSequence</h4>
<pre>public&nbsp;boolean&nbsp;isSetSequence()</pre>
<div class="block">Returns whether the attribute "sequence" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "sequence" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetSynthetase--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetSynthetase</h4>
<pre>public&nbsp;boolean&nbsp;isSetSynthetase()</pre>
<div class="block">Returns whether the attribute "synthetase" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "synthetase" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetUnprocessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetUnprocessed</h4>
<pre>public&nbsp;boolean&nbsp;isSetUnprocessed()</pre>
<div class="block">Returns whether the attribute "unprocessedProteinId" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "unprocessedProteinId" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetProcessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetProcessed</h4>
<pre>public&nbsp;boolean&nbsp;isSetProcessed()</pre>
<div class="block">Returns whether the attribute "processedProteinId" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "processedProteinId" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetComplexId--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetComplexId</h4>
<pre>public&nbsp;boolean&nbsp;isSetComplexId()</pre>
<div class="block">Returns whether the attribute "complexId" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "complexId" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetKeff--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetKeff</h4>
<pre>public&nbsp;boolean&nbsp;isSetKeff()</pre>
<div class="block">Returns whether the attribute "keff" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "keff" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetDataId--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetDataId</h4>
<pre>public&nbsp;boolean&nbsp;isSetDataId()</pre>
<div class="block">Returns whether the attribute "dataId" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "dataId" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetReverse--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetReverse</h4>
<pre>public&nbsp;boolean&nbsp;isSetReverse()</pre>
<div class="block">Returns whether the attribute "reverse" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "reverse" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetCodon--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetCodon</h4>
<pre>public&nbsp;boolean&nbsp;isSetCodon()</pre>
<div class="block">Returns whether the attribute "codon" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "codon" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetAminoAcid--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetAminoAcid</h4>
<pre>public&nbsp;boolean&nbsp;isSetAminoAcid()</pre>
<div class="block">Returns whether the attribute "aminoAcid" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "aminoAcid" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetBiomassType--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetBiomassType</h4>
<pre>public&nbsp;boolean&nbsp;isSetBiomassType()</pre>
<div class="block">Returns whether the attribute "biomassType" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "biomassType" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetSurfaceAreaInner--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetSurfaceAreaInner</h4>
<pre>public&nbsp;boolean&nbsp;isSetSurfaceAreaInner()</pre>
<div class="block">Returns whether the attribute "surfaceAreaInnerMembrane" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "surfaceAreaInnerMembrane" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetSurfaceAreaOuter--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetSurfaceAreaOuter</h4>
<pre>public&nbsp;boolean&nbsp;isSetSurfaceAreaOuter()</pre>
<div class="block">Returns whether the attribute "surfaceAreaOuterMembrane" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "surfaceAreaOuterMembrane" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetAggregationPropensity--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetAggregationPropensity</h4>
<pre>public&nbsp;boolean&nbsp;isSetAggregationPropensity()</pre>
<div class="block">Returns whether the attribute "aggregationPropensity" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "aggregationPropensity" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetPropensityScaling--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetPropensityScaling</h4>
<pre>public&nbsp;boolean&nbsp;isSetPropensityScaling()</pre>
<div class="block">Returns whether the attribute "propensityScaling" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "propensityScaling" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetVariableKind--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetVariableKind</h4>
<pre>public&nbsp;boolean&nbsp;isSetVariableKind()</pre>
<div class="block">Returns whether the attribute "variableKind" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "variableKind" is set.</dd>
</dl>
</li>
</ul>
<a name="getSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequence</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSequence()</pre>
<div class="block">Returns the value of the attribute "sequence".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "sequence".</dd>
</dl>
</li>
</ul>
<a name="getComplexId--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getComplexId</h4>
<pre>public&nbsp;java.lang.String&nbsp;getComplexId()</pre>
<div class="block">Returns the value of the attribute "complexId".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "complexId".</dd>
</dl>
</li>
</ul>
<a name="getUnprocessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getUnprocessed</h4>
<pre>public&nbsp;java.lang.String&nbsp;getUnprocessed()</pre>
<div class="block">Returns the value of the attribute "unprocessedProteinId".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "unprocessedProteinId".</dd>
</dl>
</li>
</ul>
<a name="getProcessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getProcessed</h4>
<pre>public&nbsp;java.lang.String&nbsp;getProcessed()</pre>
<div class="block">Returns the value of the attribute "processedProteinId".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "processedProteinId".</dd>
</dl>
</li>
</ul>
<a name="getSynthetase--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSynthetase</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSynthetase()</pre>
<div class="block">Returns the value of the attribute "synthetase".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "synthetase".</dd>
</dl>
</li>
</ul>
<a name="getKeff--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getKeff</h4>
<pre>public&nbsp;java.lang.String&nbsp;getKeff()</pre>
<div class="block">Returns the value of the attribute "keff".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "keff".</dd>
</dl>
</li>
</ul>
<a name="getDataId--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getDataId</h4>
<pre>public&nbsp;java.lang.String&nbsp;getDataId()</pre>
<div class="block">Returns the value of the attribute "dataId".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "dataId".</dd>
</dl>
</li>
</ul>
<a name="getReverse--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getReverse</h4>
<pre>public&nbsp;java.lang.String&nbsp;getReverse()</pre>
<div class="block">Returns the value of the attribute "reverse".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "reverse".</dd>
</dl>
</li>
</ul>
<a name="getCodon--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getCodon</h4>
<pre>public&nbsp;java.lang.String&nbsp;getCodon()</pre>
<div class="block">Returns the value of the attribute "codon".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "codon".</dd>
</dl>
</li>
</ul>
<a name="getAminoAcid--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAminoAcid</h4>
<pre>public&nbsp;java.lang.String&nbsp;getAminoAcid()</pre>
<div class="block">Returns the value of the attribute "aminoAcid".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "aminoAcid".</dd>
</dl>
</li>
</ul>
<a name="getBiomassType--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getBiomassType</h4>
<pre>public&nbsp;java.lang.String&nbsp;getBiomassType()</pre>
<div class="block">Returns the value of the attribute "biomassType".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "biomassType".</dd>
</dl>
</li>
</ul>
<a name="getSurfaceAreaInner--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSurfaceAreaInner</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSurfaceAreaInner()</pre>
<div class="block">Returns the value of the attribute "surfaceAreaInnerMembrane".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "surfaceAreaInnerMembrane".</dd>
</dl>
</li>
</ul>
<a name="getSurfaceAreaOuter--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSurfaceAreaOuter</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSurfaceAreaOuter()</pre>
<div class="block">Returns the value of the attribute "surfaceAreaOuterMembrane".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "surfaceAreaOuterMembrane".</dd>
</dl>
</li>
</ul>
<a name="getAggregationPropensity--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getAggregationPropensity</h4>
<pre>public&nbsp;java.lang.String&nbsp;getAggregationPropensity()</pre>
<div class="block">Returns the value of the attribute "aggregationPropensity".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "aggregationPropensity".</dd>
</dl>
</li>
</ul>
<a name="getPropensityScaling--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getPropensityScaling</h4>
<pre>public&nbsp;java.lang.String&nbsp;getPropensityScaling()</pre>
<div class="block">Returns the value of the attribute "propensityScaling".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "propensityScaling".</dd>
</dl>
</li>
</ul>
<a name="getVariableKind--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getVariableKind</h4>
<pre>public&nbsp;java.lang.String&nbsp;getVariableKind()</pre>
<div class="block">Returns the value of the attribute "variableKind".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "variableKind".</dd>
</dl>
</li>
</ul>
<a name="setSequence-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSequence</h4>
<pre>public&nbsp;int&nbsp;setSequence(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "sequence".
 <p>
 The optional attribute "sequence" takes a string that represents a
 valid URI for a ComponentDefinition in an SBOL document. This attribute is
 mandatory when the reaction is of type "TranscriptionReaction" or of type
 "TranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - valid URI of a ComponentDefinition in an SBOL document</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setComplexId-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setComplexId</h4>
<pre>public&nbsp;int&nbsp;setComplexId(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "complexId".
 <p>
 The optional attribute "complexId" takes a string representing the Id of
 the ComplexData object that is referred by the reaction. This attribute is
 mandatory for reactions of type "ComplexFormationReaction" and
 "MetabolicReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - id of the ComplexData object that is referred in COBRAme</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setSynthetase-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSynthetase</h4>
<pre>public&nbsp;int&nbsp;setSynthetase(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "synthetase".
 <p>
 The optional attribute "synthetase" takes a string representing the Id of a
 synthetase. This attribute is mandatory for reactions of type
 "tRNAChargingReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - id of the synthetase that synthesizes the tRNA charging</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setUnprocessed-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setUnprocessed</h4>
<pre>public&nbsp;int&nbsp;setUnprocessed(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "unprocessedProteinId".
 <p>
 The optional attribute "unprocessedProteinId" contains the id of a species
 that is the precursor of the processed protein. This attribute is mandatory
 when the reaction is of type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - the id of the precursor of the species</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setProcessed-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setProcessed</h4>
<pre>public&nbsp;int&nbsp;setProcessed(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "processedProteinId".
 <p>
 The optional attribute "processedProteinId" contains the id of the species
 that is modified by the reaction. This attribute is mandatory
 when the reaction is of type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - the id of the processed species</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setKeff-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setKeff</h4>
<pre>public&nbsp;int&nbsp;setKeff(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "keff".
 <p>
 The optional attribute "keff" contains the turnover rate of an enzyme
 catalyzing a reaction. The attribute is mandatory for reactions of type
 "MetabolicReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - the effective turnover rate of the enzyme catalyzing the reaction</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setDataId-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setDataId</h4>
<pre>public&nbsp;int&nbsp;setDataId(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "dataId".
 <p>
 The optional attribute "dataId" contains the Id of the process data object
 in COBRAme that contains additional information about the reaction. The
 attribute is mandatory for all reaction types except "SummaryVariable",
 "MEReaction" and "GenericFormationreaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - the Id of the process data object in COBRAme containing additional
        reaction information.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setReverse-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setReverse</h4>
<pre>public&nbsp;int&nbsp;setReverse(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "reverse".
 <p>
 The optional attribute "reverse" contains a boolean representing if the
 "MetabolicReaction" is the reverse direction of a reversible reaction or
 not.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - whether the reaction represents the reverse direction or not</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setCodon-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setCodon</h4>
<pre>public&nbsp;int&nbsp;setCodon(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "codon".
 <p>
 The optional attribute "codon" takes a string representing a codon that
 will be translated to a specific amino acid. This attribute should only be
 used in reactions of type "tRNAChargingReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a String containing a triplet of nucleotides in single-letter
        representation</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setAminoAcid-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setAminoAcid</h4>
<pre>public&nbsp;int&nbsp;setAminoAcid(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "aminoAcid".
 <p>
 The optional attribute "aminoAcid" takes a string representing an amino
 acid that will be added to a peptide by the charged tRNA. This attribute
 should only be used in reactions of type "tRNAChargingReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a String containing the id of an amino acid</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setBiomassType-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setBiomassType</h4>
<pre>public&nbsp;int&nbsp;setBiomassType(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "biomassType".
 <p>
 The optional attribute "biomassType" takes a string representing a type of
 biomass that is used in reactions of type "PostTranslationReaction". This
 attribute is used when biomass is added to the translated protein to
 indicate which type of biomass is added by the reaction.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a String containing the id of an constraint representing a type of
        biomass</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setSurfaceAreaInner-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSurfaceAreaInner</h4>
<pre>public&nbsp;int&nbsp;setSurfaceAreaInner(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "surfaceAreaInnerMembrane".
 <p>
 The optional attribute "surfaceAreaInnerMembrane" contains information
 about the area that is occupied by the protein altered by the reaction of
 type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a double representing the area that is occupied by the protein</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setSurfaceAreaOuter-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSurfaceAreaOuter</h4>
<pre>public&nbsp;int&nbsp;setSurfaceAreaOuter(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "surfaceAreaOuterMembrane".
 <p>
 The optional attribute "surfaceAreaOuterMembrane" contains information
 about the area that is occupied by the protein altered by the reaction of
 type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a double representing the area that is occupied by the protein</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setAggregationPropensity-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setAggregationPropensity</h4>
<pre>public&nbsp;int&nbsp;setAggregationPropensity(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "aggregationPropensity".
 <p>
 The optional attribute "aggregationPropensity" contains information
 about propensity of the protein to aggregate. This attribute is mandatory
 for reactions of type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a double representing the propensity of the protein to aggregate.</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setPropensityScaling-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setPropensityScaling</h4>
<pre>public&nbsp;int&nbsp;setPropensityScaling(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "propensityScaling".
 <p>
 The optional attribute "propensityScaling" contains information
 about propensity of the protein to aggregate. This attribute is mandatory
 for reactions of type "PostTranslationReaction".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - string representation of a double value accounting for the
        propensity of some peptides to be folded by certain chaperones</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setVariableKind-java.lang.String-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>setVariableKind</h4>
<pre>public&nbsp;int&nbsp;setVariableKind(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the mandatory attribute "variableKind".
 <p>
 The attribute defines if the values for the reaction flux are continuous or
 discrete.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - a string, either "continuous" or "discrete"</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MEReactionPlugin.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../sbmlme/MEProcessData.html" title="class in sbmlme"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../sbmlme/MESBOLPlugin.html" title="class in sbmlme"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?sbmlme/MEReactionPlugin.html" target="_top">Frames</a></li>
<li><a href="MEReactionPlugin.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
