<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="de">
<head>
<!-- Generated by javadoc (1.8.0_162) on Thu Mar 15 02:45:56 CET 2018 -->
<title>MESpeciesPlugin</title>
<meta name="date" content="2018-03-15">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
<script type="text/javascript" src="../script.js"></script>
</head>
<body>
<script type="text/javascript"><!--
    try {
        if (location.href.indexOf('is-external=true') == -1) {
            parent.document.title="MESpeciesPlugin";
        }
    }
    catch(err) {
    }
//-->
var methods = {"i0":10,"i1":10,"i2":10,"i3":10,"i4":10,"i5":10,"i6":10,"i7":10,"i8":10,"i9":10,"i10":10,"i11":10,"i12":10};
var tabs = {65535:["t0","All Methods"],2:["t2","Instance Methods"],8:["t4","Concrete Methods"]};
var altColor = "altColor";
var rowColor = "rowColor";
var tableTab = "tableTab";
var activeTableTab = "activeTableTab";
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar.top">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.top" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.top.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MESpeciesPlugin.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../sbmlme/MESBOLPlugin.html" title="class in sbmlme"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../sbmlme/MESpeciesReference.html" title="class in sbmlme"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?sbmlme/MESpeciesPlugin.html" target="_top">Frames</a></li>
<li><a href="MESpeciesPlugin.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">sbmlme</div>
<h2 title="Class MESpeciesPlugin" class="title">Class MESpeciesPlugin</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.AbstractTreeNode</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.xml.XMLToken</li>
<li>
<ul class="inheritance">
<li>org.sbml.jsbml.xml.XMLNode</li>
<li>
<ul class="inheritance">
<li><a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">sbmlme.MEAbstractXMLNodePlugin</a></li>
<li>
<ul class="inheritance">
<li>sbmlme.MESpeciesPlugin</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd>java.io.Serializable, java.lang.Cloneable, javax.swing.tree.TreeNode, org.sbml.jsbml.util.TreeNodeWithChangeSupport, <a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="typeNameLabel">MESpeciesPlugin</span>
extends <a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">MEAbstractXMLNodePlugin</a>
implements <a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></pre>
<div class="block">implements attributes and methods to encode all information of species in a
 COBRAme model in SBML.</div>
<dl>
<dt><span class="simpleTagLabel">Author:</span></dt>
<dd>Marc A. Voigt</dd>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../serialized-form.html#sbmlme.MESpeciesPlugin">Serialized Form</a></dd>
</dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>private static long</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#serialVersionUID">serialVersionUID</a></span></code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.org.sbml.jsbml.xml.XMLToken">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;org.sbml.jsbml.xml.XMLToken</h3>
<code>attributes, characters, column, isEndElement, isEOF, isStartElement, isText, line, namespaces, triple</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.org.sbml.jsbml.AbstractTreeNode">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;org.sbml.jsbml.AbstractTreeNode</h3>
<code>listOfListeners, parent, resourceBundle</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="fields.inherited.from.class.sbmlme.MEConstants">
<!--   -->
</a>
<h3>Fields inherited from interface&nbsp;sbmlme.<a href="../sbmlme/MEConstants.html" title="interface in sbmlme">MEConstants</a></h3>
<code><a href="../sbmlme/MEConstants.html#aggregationPropensity">aggregationPropensity</a>, <a href="../sbmlme/MEConstants.html#aminoAcid">aminoAcid</a>, <a href="../sbmlme/MEConstants.html#biomassType">biomassType</a>, <a href="../sbmlme/MEConstants.html#codon">codon</a>, <a href="../sbmlme/MEConstants.html#coefficient">coefficient</a>, <a href="../sbmlme/MEConstants.html#complexFormationReaction">complexFormationReaction</a>, <a href="../sbmlme/MEConstants.html#complexId">complexId</a>, <a href="../sbmlme/MEConstants.html#dataId">dataId</a>, <a href="../sbmlme/MEConstants.html#dna">dna</a>, <a href="../sbmlme/MEConstants.html#dnaSegment">dnaSegment</a>, <a href="../sbmlme/MEConstants.html#element">element</a>, <a href="../sbmlme/MEConstants.html#enzymeRef">enzymeRef</a>, <a href="../sbmlme/MEConstants.html#fixedkeff">fixedkeff</a>, <a href="../sbmlme/MEConstants.html#genericFormation">genericFormation</a>, <a href="../sbmlme/MEConstants.html#genomePos">genomePos</a>, <a href="../sbmlme/MEConstants.html#id">id</a>, <a href="../sbmlme/MEConstants.html#keff">keff</a>, <a href="../sbmlme/MEConstants.html#key">key</a>, <a href="../sbmlme/MEConstants.html#lengthDependent">lengthDependent</a>, <a href="../sbmlme/MEConstants.html#lowerFluxBound">lowerFluxBound</a>, <a href="../sbmlme/MEConstants.html#meReaction">meReaction</a>, <a href="../sbmlme/MEConstants.html#messengerRNA">messengerRNA</a>, <a href="../sbmlme/MEConstants.html#metabolicReaction">metabolicReaction</a>, <a href="../sbmlme/MEConstants.html#mRNA">mRNA</a>, <a href="../sbmlme/MEConstants.html#multiplier">multiplier</a>, <a href="../sbmlme/MEConstants.html#name">name</a>, <a href="../sbmlme/MEConstants.html#ncRNA">ncRNA</a>, <a href="../sbmlme/MEConstants.html#noncodingRNA">noncodingRNA</a>, <a href="../sbmlme/MEConstants.html#ns">ns</a>, <a href="../sbmlme/MEConstants.html#postTranslationReaction">postTranslationReaction</a>, <a href="../sbmlme/MEConstants.html#prefix">prefix</a>, <a href="../sbmlme/MEConstants.html#propensityScaling">propensityScaling</a>, <a href="../sbmlme/MEConstants.html#protein">protein</a>, <a href="../sbmlme/MEConstants.html#proteinComplex">proteinComplex</a>, <a href="../sbmlme/MEConstants.html#rate">rate</a>, <a href="../sbmlme/MEConstants.html#reactionPlugin">reactionPlugin</a>, <a href="../sbmlme/MEConstants.html#reverse">reverse</a>, <a href="../sbmlme/MEConstants.html#ribosomalRNA">ribosomalRNA</a>, <a href="../sbmlme/MEConstants.html#rRNA">rRNA</a>, <a href="../sbmlme/MEConstants.html#sbolAnnotation">sbolAnnotation</a>, <a href="../sbmlme/MEConstants.html#sbolLoc">sbolLoc</a>, <a href="../sbmlme/MEConstants.html#sbolURI">sbolURI</a>, <a href="../sbmlme/MEConstants.html#seq">seq</a>, <a href="../sbmlme/MEConstants.html#sequence">sequence</a>, <a href="../sbmlme/MEConstants.html#species">species</a>, <a href="../sbmlme/MEConstants.html#speciesPlugin">speciesPlugin</a>, <a href="../sbmlme/MEConstants.html#speciesRef">speciesRef</a>, <a href="../sbmlme/MEConstants.html#stoichiometry">stoichiometry</a>, <a href="../sbmlme/MEConstants.html#subreaction">subreaction</a>, <a href="../sbmlme/MEConstants.html#summaryVariable">summaryVariable</a>, <a href="../sbmlme/MEConstants.html#surfaceAreaInner">surfaceAreaInner</a>, <a href="../sbmlme/MEConstants.html#surfaceAreaOuter">surfaceAreaOuter</a>, <a href="../sbmlme/MEConstants.html#synthetase">synthetase</a>, <a href="../sbmlme/MEConstants.html#temperature">temperature</a>, <a href="../sbmlme/MEConstants.html#transcription">transcription</a>, <a href="../sbmlme/MEConstants.html#transferRNA">transferRNA</a>, <a href="../sbmlme/MEConstants.html#translation">translation</a>, <a href="../sbmlme/MEConstants.html#translocation">translocation</a>, <a href="../sbmlme/MEConstants.html#tRNA">tRNA</a>, <a href="../sbmlme/MEConstants.html#tRNACharging">tRNACharging</a>, <a href="../sbmlme/MEConstants.html#unprocessed">unprocessed</a>, <a href="../sbmlme/MEConstants.html#upperFluxBound">upperFluxBound</a>, <a href="../sbmlme/MEConstants.html#value">value</a>, <a href="../sbmlme/MEConstants.html#variableKind">variableKind</a>, <a href="../sbmlme/MEConstants.html#versionOne">versionOne</a></code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#MESpeciesPlugin--">MESpeciesPlugin</a></span>()</code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colOne"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#MESpeciesPlugin-sbmlme.MESpeciesPlugin-">MESpeciesPlugin</a></span>(<a href="../sbmlme/MESpeciesPlugin.html" title="class in sbmlme">MESpeciesPlugin</a>&nbsp;sp)</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method.summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="memberSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span id="t0" class="activeTableTab"><span>All Methods</span><span class="tabEnd">&nbsp;</span></span><span id="t2" class="tableTab"><span><a href="javascript:show(2);">Instance Methods</a></span><span class="tabEnd">&nbsp;</span></span><span id="t4" class="tableTab"><span><a href="javascript:show(8);">Concrete Methods</a></span><span class="tabEnd">&nbsp;</span></span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr id="i0" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#createMEProcessedSpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">createMEProcessedSpecies</a></span>(org.sbml.jsbml.Model&nbsp;model,
                        java.lang.String&nbsp;id,
                        java.lang.String&nbsp;formula,
                        java.lang.String&nbsp;name,
                        java.lang.String&nbsp;compartment,
                        java.lang.String&nbsp;unprocessedId,
                        org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                        java.lang.String&nbsp;groupName)</code>
<div class="block">Create and add an SBMLme species to the model that represents a processed
 protein.</div>
</td>
</tr>
<tr id="i1" class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#createMESequenceSpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbolstandard.core2.SBOLDocument-java.lang.String-java.lang.String-java.lang.String-int-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">createMESequenceSpecies</a></span>(org.sbml.jsbml.Model&nbsp;model,
                       java.lang.String&nbsp;id,
                       java.lang.String&nbsp;formula,
                       java.lang.String&nbsp;name,
                       java.lang.String&nbsp;compartment,
                       org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                       java.lang.String&nbsp;sequence,
                       java.lang.String&nbsp;role,
                       java.lang.String&nbsp;orientation,
                       int&nbsp;genomePosition,
                       org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                       java.lang.String&nbsp;groupName)</code>
<div class="block">Create and add an SBMLme species to the model that represents a RNA.</div>
</td>
</tr>
<tr id="i2" class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#createMESpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">createMESpecies</a></span>(org.sbml.jsbml.Model&nbsp;model,
               java.lang.String&nbsp;id,
               java.lang.String&nbsp;formula,
               java.lang.String&nbsp;name,
               java.lang.String&nbsp;compartment,
               org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
               java.lang.String&nbsp;groupName)</code>
<div class="block">Create and add an SBMLme species to the model that does not require
 additional attributes.</div>
</td>
</tr>
<tr id="i3" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#getGenomePos--">getGenomePos</a></span>()</code>
<div class="block">Returns the value of the attribute "genomePosition".</div>
</td>
</tr>
<tr id="i4" class="altColor">
<td class="colFirst"><code>org.sbml.jsbml.Compartment</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#getSBMLCompartment-java.lang.String-org.sbml.jsbml.Model-">getSBMLCompartment</a></span>(java.lang.String&nbsp;compartment,
                  org.sbml.jsbml.Model&nbsp;model)</code>
<div class="block">get the Compartment in the SBML model for the given string, create a
 Compartment if not already present</div>
</td>
</tr>
<tr id="i5" class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#getSequence--">getSequence</a></span>()</code>
<div class="block">Returns the value of the attribute "sequence".</div>
</td>
</tr>
<tr id="i6" class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#getUnprocessed--">getUnprocessed</a></span>()</code>
<div class="block">Returns the value of the attribute "unprocessedProteinId".</div>
</td>
</tr>
<tr id="i7" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#isSetGenomePos--">isSetGenomePos</a></span>()</code>
<div class="block">Returns whether the attribute "genomePosition" is set.</div>
</td>
</tr>
<tr id="i8" class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#isSetSequence--">isSetSequence</a></span>()</code>
<div class="block">Returns whether the attribute "sequence" is set.</div>
</td>
</tr>
<tr id="i9" class="rowColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#isSetUnprocessed--">isSetUnprocessed</a></span>()</code>
<div class="block">Returns whether the attribute "unprocessedProteinId" is set.</div>
</td>
</tr>
<tr id="i10" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#setGenomePos-java.lang.String-">setGenomePos</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "genomePosition".</div>
</td>
</tr>
<tr id="i11" class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#setSequence-java.lang.String-">setSequence</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "sequence".</div>
</td>
</tr>
<tr id="i12" class="altColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><span class="memberNameLink"><a href="../sbmlme/MESpeciesPlugin.html#setUnprocessed-java.lang.String-">setUnprocessed</a></span>(java.lang.String&nbsp;value)</code>
<div class="block">Sets the value of the attribute "unprocessedProteinId".</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.sbmlme.MEAbstractXMLNodePlugin">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;sbmlme.<a href="../sbmlme/MEAbstractXMLNodePlugin.html" title="class in sbmlme">MEAbstractXMLNodePlugin</a></h3>
<code><a href="../sbmlme/MEAbstractXMLNodePlugin.html#createSBMLConformId-java.lang.String-">createSBMLConformId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getAttribute-java.lang.String-">getAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getId--">getId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#getName--">getName</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetAttribute-java.lang.String-">isSetAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetId--">isSetId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#isSetName--">isSetName</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setAttribute-java.lang.String-java.lang.String-">setAttribute</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setId-java.lang.String-">setId</a>, <a href="../sbmlme/MEAbstractXMLNodePlugin.html#setName-java.lang.String-">setName</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.xml.XMLNode">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.xml.XMLNode</h3>
<code>addChild, clone, convertStringToXMLNode, convertStringToXMLNode, convertXMLNodeToString, getAllowsChildren, getChild, getChildAt, getChildCount, getChildElement, getChildElements, insertChild, removeChild, removeChild, removeChildren, toString, toXMLString</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.xml.XMLToken">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.xml.XMLToken</h3>
<code>addAttr, addAttr, addAttr, addAttr, addNamespace, addNamespace, append, clearAttributes, clearNamespaces, equals, getAttributes, getAttributesLength, getAttrIndex, getAttrIndex, getAttrIndex, getAttrName, getAttrPrefix, getAttrPrefixedName, getAttrURI, getAttrValue, getAttrValue, getAttrValue, getAttrValue, getCharacters, getColumn, getLine, getNamespaceIndex, getNamespaceIndexByPrefix, getNamespacePrefix, getNamespacePrefix, getNamespaces, getNamespacesLength, getNamespaceURI, getNamespaceURI, getNamespaceURI, getPrefix, getURI, hasAttr, hasAttr, hasAttr, hasAttr, hashCode, hasNamespaceNS, hasNamespacePrefix, hasNamespaceURI, isAttributesEmpty, isElement, isEnd, isEndFor, isEOF, isNamespacesEmpty, isStart, isText, removeAttr, removeAttr, removeAttr, removeAttr, removeNamespace, removeNamespace, setAttributes, setCharacters, setEnd, setEOF, setNamespaces, setTriple, unsetEnd</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.org.sbml.jsbml.AbstractTreeNode">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;org.sbml.jsbml.AbstractTreeNode</h3>
<code>addAllChangeListeners, addAllChangeListeners, addTreeNodeChangeListener, addTreeNodeChangeListener, children, clearUserObjects, containsUserObjectKey, filter, filter, filter, fireNodeAddedEvent, fireNodeRemovedEvent, firePropertyChange, getIndex, getListOfTreeNodeChangeListeners, getNumChildren, getParent, getRoot, getTreeNodeChangeListenerCount, getUserObject, indexOf, isCloningInProgress, isInvalidSBMLAllowed, isLeaf, isReadingInProgress, isRoot, isSetParent, isSetUserObjects, notifyChildChange, putUserObject, removeAllTreeNodeChangeListeners, removeAllTreeNodeChangeListeners, removeFromParent, removeTreeNodeChangeListener, removeTreeNodeChangeListener, removeUserObject, setParent, userObjectKeySet</code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods.inherited.from.class.java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>finalize, getClass, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field.detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="serialVersionUID">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>serialVersionUID</h4>
<pre>private static final&nbsp;long serialVersionUID</pre>
<dl>
<dt><span class="seeLabel">See Also:</span></dt>
<dd><a href="../constant-values.html#sbmlme.MESpeciesPlugin.serialVersionUID">Constant Field Values</a></dd>
</dl>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor.detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="MESpeciesPlugin--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MESpeciesPlugin</h4>
<pre>public&nbsp;MESpeciesPlugin()</pre>
</li>
</ul>
<a name="MESpeciesPlugin-sbmlme.MESpeciesPlugin-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>MESpeciesPlugin</h4>
<pre>public&nbsp;MESpeciesPlugin(<a href="../sbmlme/MESpeciesPlugin.html" title="class in sbmlme">MESpeciesPlugin</a>&nbsp;sp)</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method.detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="getSBMLCompartment-java.lang.String-org.sbml.jsbml.Model-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSBMLCompartment</h4>
<pre>public&nbsp;org.sbml.jsbml.Compartment&nbsp;getSBMLCompartment(java.lang.String&nbsp;compartment,
                                                     org.sbml.jsbml.Model&nbsp;model)</pre>
<div class="block">get the Compartment in the SBML model for the given string, create a
 Compartment if not already present</div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>compartment</code> - the id of the desired compartment</dd>
<dd><code>model</code> - the SBML model</dd>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the Compartment with the given id</dd>
</dl>
</li>
</ul>
<a name="createMESpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createMESpecies</h4>
<pre>public&nbsp;void&nbsp;createMESpecies(org.sbml.jsbml.Model&nbsp;model,
                            java.lang.String&nbsp;id,
                            java.lang.String&nbsp;formula,
                            java.lang.String&nbsp;name,
                            java.lang.String&nbsp;compartment,
                            org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                            java.lang.String&nbsp;groupName)</pre>
<div class="block">Create and add an SBMLme species to the model that does not require
 additional attributes.
 <p>
 This method should be used for all COBRAme species except for species of
 type "TranscribedGene" or "ProcessedProtein". These two types of species
 require additional attributes which cannot be encoded by SBML or one of its
 packages and should be added to the model by their own methods.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>id</code> - the unique id of the species to be created</dd>
<dd><code>formula</code> - the formula of the species, can be null</dd>
<dd><code>name</code> - the name of the species, can be null</dd>
<dd><code>compartment</code> - the id of the compartment</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>groupName</code> - the id of the group that the species should be a part of. This
        should equal to the type of species in the COBRAme model.</dd>
</dl>
</li>
</ul>
<a name="createMEProcessedSpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createMEProcessedSpecies</h4>
<pre>public&nbsp;void&nbsp;createMEProcessedSpecies(org.sbml.jsbml.Model&nbsp;model,
                                     java.lang.String&nbsp;id,
                                     java.lang.String&nbsp;formula,
                                     java.lang.String&nbsp;name,
                                     java.lang.String&nbsp;compartment,
                                     java.lang.String&nbsp;unprocessedId,
                                     org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                     java.lang.String&nbsp;groupName)</pre>
<div class="block">Create and add an SBMLme species to the model that represents a processed
 protein.
 <p>
 This method should be used for COBRAme species of type "ProcessedProtein".
 This method adds an additional attribute named "unprocessedProteinId" to
 the annotation of the species which contains the id of the precursor of the
 protein.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>id</code> - the unique id of the species to be created</dd>
<dd><code>formula</code> - the formula of the species, can be null</dd>
<dd><code>name</code> - the name of the species, can be null</dd>
<dd><code>compartment</code> - the id of the compartment</dd>
<dd><code>unprocessedId</code> - the id of the unprocessed protein that is the precursor of the
        processed species created by this method</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>groupName</code> - the id of the group that the species should be a part of. This
        should equal to the type of species in the COBRAme model.</dd>
</dl>
</li>
</ul>
<a name="createMESequenceSpecies-org.sbml.jsbml.Model-java.lang.String-java.lang.String-java.lang.String-java.lang.String-org.sbolstandard.core2.SBOLDocument-java.lang.String-java.lang.String-java.lang.String-int-org.sbml.jsbml.ext.groups.GroupsModelPlugin-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>createMESequenceSpecies</h4>
<pre>public&nbsp;void&nbsp;createMESequenceSpecies(org.sbml.jsbml.Model&nbsp;model,
                                    java.lang.String&nbsp;id,
                                    java.lang.String&nbsp;formula,
                                    java.lang.String&nbsp;name,
                                    java.lang.String&nbsp;compartment,
                                    org.sbolstandard.core2.SBOLDocument&nbsp;sbol,
                                    java.lang.String&nbsp;sequence,
                                    java.lang.String&nbsp;role,
                                    java.lang.String&nbsp;orientation,
                                    int&nbsp;genomePosition,
                                    org.sbml.jsbml.ext.groups.GroupsModelPlugin&nbsp;groups,
                                    java.lang.String&nbsp;groupName)
                             throws org.sbolstandard.core2.SBOLValidationException</pre>
<div class="block">Create and add an SBMLme species to the model that represents a RNA.
 <p>
 This method should be used for COBRAme species of type "TranscribedGene".
 The method adds the nucleotide sequence of the species with information
 regarding its location to the given SBOL document. The method adds two
 attributes to the annotation of the species in SBML, "sequence" and
 "genomePosition". The attribute "sequence" contains the URI of the
 ComponentDefinition in the SBOL document the nucleotide has been added to.
 The attribute "genomePosition" contains the starting position of the
 nucleotide sequence on the genome of the modeled organism. The attribute
 "genomePosition" is needed since SBOL defines position in relation to a
 sequence while COBRAme defines position in relation to a genome.
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>model</code> - the SBML model</dd>
<dd><code>id</code> - the unique id of the species to be created</dd>
<dd><code>formula</code> - the formula of the species, can be null</dd>
<dd><code>name</code> - the name of the species, can be null</dd>
<dd><code>compartment</code> - the id of the compartment</dd>
<dd><code>sbol</code> - the SBOL document to which the sequence should be added</dd>
<dd><code>sequence</code> - the nucleotide sequence of the species</dd>
<dd><code>role</code> - the type of RNA the species belongs to</dd>
<dd><code>orientation</code> - the strand on the genome that the sequence is located on</dd>
<dd><code>genomePosition</code> - the position on the genome where the sequence starts</dd>
<dd><code>groups</code> - the GroupsModelPlugin of the SBML model</dd>
<dd><code>groupName</code> - the id of the group that the species should be a part of. This
        should equal to the type of species in the COBRAme model.</dd>
<dt><span class="throwsLabel">Throws:</span></dt>
<dd><code>org.sbolstandard.core2.SBOLValidationException</code></dd>
</dl>
</li>
</ul>
<a name="isSetSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetSequence</h4>
<pre>public&nbsp;boolean&nbsp;isSetSequence()</pre>
<div class="block">Returns whether the attribute "sequence" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "sequence" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetGenomePos--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetGenomePos</h4>
<pre>public&nbsp;boolean&nbsp;isSetGenomePos()</pre>
<div class="block">Returns whether the attribute "genomePosition" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "genomePosition" is set.</dd>
</dl>
</li>
</ul>
<a name="isSetUnprocessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isSetUnprocessed</h4>
<pre>public&nbsp;boolean&nbsp;isSetUnprocessed()</pre>
<div class="block">Returns whether the attribute "unprocessedProteinId" is set.</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>whether the attribute "unprocessedProteinId" is set.</dd>
</dl>
</li>
</ul>
<a name="getSequence--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getSequence</h4>
<pre>public&nbsp;java.lang.String&nbsp;getSequence()</pre>
<div class="block">Returns the value of the attribute "sequence".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "sequence".</dd>
</dl>
</li>
</ul>
<a name="getGenomePos--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenomePos</h4>
<pre>public&nbsp;java.lang.String&nbsp;getGenomePos()</pre>
<div class="block">Returns the value of the attribute "genomePosition".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "genomePosition".</dd>
</dl>
</li>
</ul>
<a name="getUnprocessed--">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getUnprocessed</h4>
<pre>public&nbsp;java.lang.String&nbsp;getUnprocessed()</pre>
<div class="block">Returns the value of the attribute "unprocessedProteinId".</div>
<dl>
<dt><span class="returnLabel">Returns:</span></dt>
<dd>the value of the attribute "unprocessedProteinId".</dd>
</dl>
</li>
</ul>
<a name="setSequence-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setSequence</h4>
<pre>public&nbsp;int&nbsp;setSequence(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "sequence".
 <p>
 The optional attribute "sequence" takes a string that represents a
 valid URI for a ComponentDefinition in an SBOL document. This attribute is
 mandatory when the species is of type "TranscribedGene".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - valid URI of a ComponentDefinition in an SBOL document</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setGenomePos-java.lang.String-">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setGenomePos</h4>
<pre>public&nbsp;int&nbsp;setGenomePos(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "genomePosition".
 <p>
 The optional attribute "genomePosition" contains the position on the genome
 where the nucleotide sequence starts. This attribute is mandatory when the
 species is of type "TranscribedGene".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - position on the genome where the nucleotide sequence starts</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
<a name="setUnprocessed-java.lang.String-">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>setUnprocessed</h4>
<pre>public&nbsp;int&nbsp;setUnprocessed(java.lang.String&nbsp;value)</pre>
<div class="block">Sets the value of the attribute "unprocessedProteinId".
 <p>
 The optional attribute "unprocessedProteinId" contains the id of a species
 that is the precursor of the processed protein. This attribute is mandatory
 when the species is of type "ProcessedProtein".
 </p></div>
<dl>
<dt><span class="paramLabel">Parameters:</span></dt>
<dd><code>value</code> - the id of the precursor of the species</dd>
<dt><span class="returnLabel">Returns:</span></dt>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar.bottom">
<!--   -->
</a>
<div class="skipNav"><a href="#skip.navbar.bottom" title="Skip navigation links">Skip navigation links</a></div>
<a name="navbar.bottom.firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/MESpeciesPlugin.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../sbmlme/MESBOLPlugin.html" title="class in sbmlme"><span class="typeNameLink">Prev&nbsp;Class</span></a></li>
<li><a href="../sbmlme/MESpeciesReference.html" title="class in sbmlme"><span class="typeNameLink">Next&nbsp;Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../index.html?sbmlme/MESpeciesPlugin.html" target="_top">Frames</a></li>
<li><a href="MESpeciesPlugin.html" target="_top">No&nbsp;Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All&nbsp;Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field.summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field.detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor.detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method.detail">Method</a></li>
</ul>
</div>
<a name="skip.navbar.bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
